services/src/main/java/org/keycloak/theme/DefaultThemeManager.java

```diff
@@ -276,6 +280,53 @@ private Map<Locale, Properties> getMessagesByLocale(String baseBundlename, Local
             }
         }
 
+        protected void addlocaleTranslations(Locale locale, Properties m) throws IOException {
+            for (String l : getProperties().getProperty("locales", "").split(",")) {
+                l = l.trim();
+                String key = "locale_" + l;
+                String label = m.getProperty(key);
+                if (label != null) {
+                    continue;
+                }
+                String rl = l;
+                // This is mapping old locale codes to the new locale codes for Simplified and Traditional Chinese.
+                // Once the existing locales have been moved, this code can be removed.
+                if (l.equals("zh-CN")) {
+                    rl = "zh-HANS";
+                } else if (l.equals("zh-TW")) {
+                    rl = "zh-HANT";
+                }
+                Locale loc = Locale.forLanguageTag(rl);
+                label = capitalize(loc.getDisplayName(locale), locale);
+                if (!Objects.equals(loc, locale)) {
+                    label += " (" + capitalize(loc.getDisplayName(loc), loc) + ")";
+                }
+                m.put(key, label);
+            }
+        }
+
+        private static final Pattern LATIN_CHARACTERS;
+
+        static {
+            Pattern p;
+            try {
+                p = Pattern.compile("(\\p{L1}|\\p{InLATIN_EXTENDED_A}|\\p{InLATIN_EXTENDED_B}|\\p{InLATIN_EXTENDED_C}|\\p{InLATIN_EXTENDED_D}|\\p{InLATIN_EXTENDED_E}).*");
+            } catch (PatternSyntaxException ex) {
+                log.warn("unable to create regex for latin characters", ex);
+                // just in case the JVM doesn't recognize the language patterns used above
+                p = Pattern.compile("[a-zA-Z]");
+            }
+            LATIN_CHARACTERS = p;
+        }
+
+        private String capitalize(String name, Locale locale) {
```

Comment:
I was thinking, we already include Apache Commons right? It seems that it offers a [`StringUtils.capitalize()`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#capitalize-java.lang.String-), is that something that could be used here? Note that I am not sure if it handles the edge cases correctly that were handling here.

Reply:
Done - the dependencies is used in some parts of Keycloak, but not here. So I now added it here as well.

Reply:
Hm, reverted it as that method doesn't allow passing in a locale, and making things uppercase is locale dependent. This might need more investigation.